coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala

```diff
@@ -389,64 +332,127 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
       (part, rangeOpt.get)
     }
   }
 
+  // FIXME: this is a copy-paste of a method in the ShardKeyRegexPlanner --
+  //  more evidence that these two classes should be merged.
+  private def canSupportMultiPartitionCalls(execPlans: Seq[ExecPlan]): Boolean =
+    execPlans.forall {
+      case _: PromQlRemoteExec => false
+      case _ => true
+    }
+
+  // FIXME: this is a near-exact copy-paste of a method in the ShardKeyRegexPlanner --
+  //  more evidence that these two classes should be merged.
+  private def materializeAggregate(aggregate: Aggregate, queryContext: QueryContext): PlanResult = {
+    val plan = if (LogicalPlanUtils.hasDescendantAggregateOrJoin(aggregate.vectors)) {
+      val childPlan = materialize(aggregate.vectors, queryContext)
+      addAggregator(aggregate, queryContext, PlanResult(Seq(childPlan)))
+    } else {
+      val queryParams = queryContext.origQueryParams.asInstanceOf[PromQlQueryParams]
+      val (partitions, _, _, _) = resolvePartitionsAndRoutingKeys(aggregate, queryParams)
+      val childQueryContext = queryContext.copy(
+        plannerParams = queryContext.plannerParams.copy(skipAggregatePresent = true))
+      val execPlans = partitions.map(p => materializeForPartition(aggregate, p, childQueryContext))
+      val exec = if (execPlans.size == 1) execPlans.head
+      else {
+        if ((aggregate.operator.equals(AggregationOperator.TopK)
+          || aggregate.operator.equals(AggregationOperator.BottomK)
+          || aggregate.operator.equals(AggregationOperator.CountValues)
+          ) && !canSupportMultiPartitionCalls(execPlans))
+          throw new UnsupportedOperationException(s"Shard Key regex not supported for ${aggregate.operator}")
+        else {
+          val reducer = MultiPartitionReduceAggregateExec(queryContext, inProcessPlanDispatcher,
+            execPlans.sortWith((x, _) => !x.isInstanceOf[PromQlRemoteExec]), aggregate.operator, aggregate.params)
+          reducer.addRangeVectorTransformer(AggregatePresenter(aggregate.operator, aggregate.params,
+            RangeParams(queryParams.startSecs, queryParams.stepSecs, queryParams.endSecs)))
+          reducer
+        }
+      }
+      exec
+    }
+    PlanResult(Seq(plan))
+  }
+
   /**
    * Materialize any plan whose materialization strategy is governed by whether-or-not it
    *   contains leaves that individually span partitions.
    */
   private def materializePlanHandleSplitLeaf(logicalPlan: LogicalPlan,
                                              qContext: QueryContext): PlanResult = {
     val qParams = qContext.origQueryParams.asInstanceOf[PromQlQueryParams]
-    val hasMultiPartitionLeaves = LogicalPlan.findLeafLogicalPlans(logicalPlan)
-                                             .exists(getPartitions(_, qParams).size > 1)
+    // Create one plan per RawSeries/shard-key pair, then resolve its partitions.
+    // If any resides on more than one partition, the leaf is "split".
+    val hasMultiPartitionLeaves =
```

Comment:
This is confusing because it should be renamed to `hasSplitPartitionLeaves`. "Multi-partition" implied "split" when shard-keys were filtered only by equality.

Detailed explanation for completeness: The old logic was
```
    val hasMultiPartitionLeaves = LogicalPlan.findLeafLogicalPlans(logicalPlan)
                                             .exists(getPartitions(_, qParams).size > 1)
```
Suppose some plan has regex shard-key filters. If the resolved keys reside in multiple partitions, the `getPartitions.size` step might return some value `> 1`. The old logic would interpret that as a partition split even though none might exist.